File: benchmarks/tasks/matrix_mult/matrix_mult.ts
// SPDX-License-Identifier: MIT-0

function matMult(a: number[][], b: number[][]): number[][] {
    const n = a.length;
    const result: number[][] = Array.from({ length: n }, () => Array(n).fill(0));
    for (let i = 0; i < n; i++)
        for (let j = 0; j < n; j++)
            for (let k = 0; k < n; k++)
                result[i][j] += a[i][k] * b[k][j];
    return result;
}

const a = [[1, 2, 3], [4, 5, 6], [7, 8, 9]];
const b = [[1, 2, 3], [4, 5, 6], [7, 8, 9]];
const result = matMult(a, b);
console.log(result.flat().join(" "));

File: benchmarks/tasks/matrix_mult/matrix_mult.hs
-- SPDX-License-Identifier: MIT-0

matMul :: [[Int]] -> [[Int]] -> [[Int]]
matMul a b =
  let n = length a
      col j = map (!! j) b
      dot r c = sum (zipWith (*) r c)
  in [[dot (a !! i) (col j) | j <- [0..n-1]] | i <- [0..n-1]]

main :: IO ()
main =
  let a = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
      b = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
      result = matMul a b
      flat = concatMap id result
  in putStrLn (unwords (map show flat))
